hera_qm/xrfi.py

```diff
@@ -607,11 +685,215 @@ def xrfi_h1c_pipe(uv, Kt=8, Kf=8, sig_init=6., sig_adj=2., px_threshold=0.2,
+    alg = 'detrend_meanfilt'  # Change to meanfilt because it can mask flagged pixels
+    # Calculate metric on abscal data
+    uvf_ag2, uvf_agf2 = xrfi_pipe(uvc_a, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                  cal_mode='gain', sig_init=sig_init, sig_adj=sig_adj)
+    uvf_ax2, uvf_axf2 = xrfi_pipe(uvc_a, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                  cal_mode='tot_chisq', sig_init=sig_init, sig_adj=sig_adj)
+
+    # Calculate metric on omnical data
+    uvf_og2, uvf_ogf2 = xrfi_pipe(uvc_o, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                  cal_mode='gain', sig_init=sig_init, sig_adj=sig_adj)
+    uvf_ox2, uvf_oxf2 = xrfi_pipe(uvc_o, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                  cal_mode='tot_chisq', sig_init=sig_init, sig_adj=sig_adj)
+
+    # Calculate metric on model vis
+    uvf_v2, uvf_vf2 = xrfi_pipe(uv_v, alg=alg, xants=[], Kt=kt_size, Kf=kf_size,
+                                sig_init=sig_init, sig_adj=sig_adj)
+
+    # Calculate metric on data file
+    uvf_d2, uvf_df2 = xrfi_pipe(uv_d, alg=alg, xants=[], Kt=kt_size, Kf=kf_size,
+                                sig_init=sig_init, sig_adj=sig_adj)
+
+    # Combine the metrics together
+    uvf_metrics2 = uvf_d2.combine_metrics([uvf_og2, uvf_ox2, uvf_ag2, uvf_ax2, uvf_v2, uvf_d2],
+                                          method='quadmean', inplace=False)
+    alg_func = algorithm_dict[alg]
+    uvf_metrics2.metric_array[:, :, 0] = alg_func(uvf_metrics2.metric_array[:, :, 0],
+                                                  flags=uvf_init.flag_array[:, :, 0],
+                                                  Kt=kt_size, Kf=kf_size)
+
+    # Flag on combined metrics
+    uvf_f2 = flag(uvf_metrics2, nsig_p=sig_init)
+    uvf_fws2 = watershed_flag(uvf_metrics2, uvf_f2, nsig_p=sig_adj, inplace=False)
+    uvf_combined2 = (uvf_fws2 | uvf_ogf2 | uvf_oxf2 | uvf_agf2 | uvf_axf2
+                     | uvf_vf2 | uvf_df2 | uvf_init)
+
+    # Threshold
+    uvf_temp = uvf_combined2.copy()
+    uvf_temp.to_metric(convert_wgts=True)
+    uvf_final = flag(uvf_temp, nsig_p=1.0, nsig_f=freq_threshold, nsig_t=time_threshold)
+
+    # Write out final metrics and flags
+    outfile = '.'.join([basename, final_metrics_ext])
+    outpath = os.path.join(dirname, outfile)
+    uvf_metrics2.write(outpath, clobber=True)
+    outfile = '.'.join([basename, final_flags_ext])
+    outpath = os.path.join(dirname, outfile)
+    uvf_final.write(outpath, clobber=True)
+
+    # Save calfits with new flags
+    flag_apply(uvf_final, uvc_a, force_pol=True, history=history)
+    basename = qm_utils.strip_extension(os.path.basename(acalfits_file))
+    basename = qm_utils.strip_extension(basename)  # Also get rid of .abs
+    outfile = '.'.join([basename, cal_ext, 'calfits'])
+    outpath = os.path.join(dirname, outfile)
+    uvc_a.write_calfits(outpath, clobber=True)
+
```

Comment:
Just wanted to say, even though this function is still quite long, you've done some really great work getting it to the point where it's really truly modular. What an improvement!

Reply:
Thanks!